### PR TITLE
Add new table style for Quantum Synchrotron photon emission

### PR DESCRIPTION
**Pull request by lucafedeli88**
### Description:
This PR adds a new lookup table style for Quantum Synchrotron (QS) photon emission, called "tail-optimized QS table". The difference with respect to a standard table concerns the spacing of the points on the chi_phot/chi_part (frac) axis.  The tail-optimized lookup table allows choosing a given frac_switch. Frac points are logarithmically spaced below frac_switch and linearly spaced above  frac_switch. This allows for more points close to the high-energy tail of the energy distribution of the emitted photons.
This feature has been implemented in a new QS photon emission lookup table : tailopt_photon_emission_lookup_table_params . Internally, this lookup table uses a new 2D generic table, allowing to arbitrarily choose the mapping functions to be used for the table axes. The mapping functions are functors defined in `quantum_sync_engine_tables_detail.hpp`

### List of changes:
- A new generic 2D lookup table has been added to `picsar_tables.hpp`
- The new tail-optimized lookup table has been added to `quantum_sync_engine_tables.hpp`
- Functors used for the new table style have been added to `quantum_sync_engine_tables_detail.hpp`
- Lookup table generation methods for the new table style have been added to `quantum_sync_engine_tables_generator.hpp `
- In order to avoid excessive code duplication, routines actually implementing QS lookup tables generation have been moved, as free functions, to the new source file `quantum_sync_engine_tables_generator_detail.hpp` . These routines are called by the member functions of the lookup table classes (in `quantum_sync_engine_tables_generator.hpp `).
- The member function `aux_generate_double` of QS tables has been removed. It was used to force calculations of the lookup tables in double precision. Now, this is done directly in the table generation functions.
- Tests of the new picsar tables have been added to `test_picsar_tables.cpp
- Tests of the new QS lookup tables have been added to `test_picsar_quantum_sync_tables.cpp` and `test_picsar_quantum_sync_tables_generator.cpp`

### Notes:

Another PR will follow to expose the new table style via python 🐍 bindings

⚠️ **Warning!** This PR depends on https://github.com/ECP-WarpX/picsar/pull/33

### CODE REVIEW

**Comment by NeilZaim** [comment]
Hi @lucafedeli88. Could you fix the conflicts so that we can see if the automated tests still pass?

**Comment by NeilZaim** [comment]
In case you had not seen @lucafedeli88 we have in the automated tests errors of the type `invalid initialization of reference of type ‘const std::vector<double>&’ from expression of type ‘KokkosVectorWrapper<double>’`

**Comment by lucafedeli88** [comment]
> In case you had not seen @lucafedeli88 we have in the automated tests errors of the type `invalid initialization of reference of type ‘const std::vector<double>&’ from expression of type ‘KokkosVectorWrapper<double>’`

Thanks! It should be fixed now...

**Comment by lucafedeli88** [comment]
> > In case you had not seen @lucafedeli88 we have in the automated tests errors of the type `invalid initialization of reference of type ‘const std::vector<double>&’ from expression of type ‘KokkosVectorWrapper<double>’`
> 
> Thanks! It should be fixed now...

Weird... I now have another error... I will investigate...

**Comment by lucafedeli88** [comment]
The issue is a buffer overflow (probably a corner case of the new table style). Automated tests have sanitizers enabled, so they are able to catch the bug...

**Comment by lucafedeli88** [comment]
The issue should be fixed now. There was an error in a test case.

File: multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp [hunk 1]
```diff
@@ -895,6 +895,485 @@ namespace details{
...
+            if (raw_data.size() < min_size)
+                throw std::runtime_error("Binary data is too small \
+                to be a 2D table.");
+
+            auto it_raw_data = raw_data.begin();
+
+            if (serialization::get_out<char>(it_raw_data) !=
+                static_cast<char>(sizeof(RealType))){
+                throw std::runtime_error("Mismatch between RealType \
+                used to write and to read the 1D table");
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp]:
```suggestion
                used to write and to read the 2D table");
```
Looks like the same typo is also there line 463.

File: multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp [hunk 2]
```diff
@@ -895,6 +895,485 @@ namespace details{
...
+
+            m_how_many_x =
+                serialization::get_out<decltype(m_how_many_x)>(it_raw_data);
+            m_how_many_y =
+                serialization::get_out<decltype(m_how_many_y)>(it_raw_data);
+            if(m_how_many_x <= 0)
+                throw std::runtime_error("raw_data contains invalid data.");
+            if(m_how_many_y <= 0)
+                throw std::runtime_error("raw_data contains invalid data.");
+            m_values = VectorType(static_cast<int>(m_how_many_x*m_how_many_y));
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp:991]: Why is there a `static_cast` here? Both `m_how_many_x` and `m_how_many_y` are already `int`, no?

File: multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp [hunk 3]
```diff
@@ -895,6 +895,485 @@ namespace details{
...
+            auto vals = serialization::get_n_out<RealType>(
+                    it_raw_data,
+                    m_how_many_x*m_how_many_y);
+            std::copy(vals.begin(), vals.end(), m_values.begin());
+
+            m_x_map_functor = XMapFunctor::deserialize(it_raw_data);
+            m_y_map_functor = YMapFunctor::deserialize(it_raw_data);
+            m_ix_map_functor = IXMapFunctor::deserialize(it_raw_data);
+            m_iy_map_functor = IYMapFunctor::deserialize(it_raw_data);
+        }
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp:1005]: Does it need a call to `details::aux_sync_vec(m_values);` here? (same as in line 493)

File: multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp [hunk 4]
```diff
@@ -895,6 +895,485 @@ namespace details{
...
+        * @return the size along y
+        */
+        PXRMP_GPU_QUALIFIER PXRMP_FORCE_INLINE
+        RealType get_y_size() const noexcept
+        {
+            return get_y_max() - get_y_min();
+        }
+
+        /**
+        * Returns the size of the steps along y
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp]:
```suggestion
        * Returns the size of the steps along y between j and j+1
```

File: multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp [hunk 5]
```diff
@@ -895,6 +895,485 @@ namespace details{
...
+        * Warning: no safety check on i and j is performed!
+        *
+        * @param[in] i index of the value to be overwritten
+        * @param[in] j index of the value to be overwritten
+        * @param[in] what value to be written at position (i,j)
+        */
+        PXRMP_FORCE_INLINE
+        void set_val(const int i, const int j, const RealType what)
+        {
+            m_values[idx(i, j)] = what;
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp:1313]: Does it need a call to `details::aux_sync_vec(m_values);`?

File: multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp [hunk 6]
```diff
@@ -895,6 +895,485 @@ namespace details{
...
+        * Warning: no safety check on i is performed!
+        *
+        * @param[in] i index of the value to be overwritten
+        * @param[in] what value to be written at position i
+        */
+        PXRMP_FORCE_INLINE
+        void set_val(const int i, const RealType what)
+        {
+            m_values[i] = what;
+        }
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp:1331]: Does it need a call to details::aux_sync_vec(m_values);?

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables.hpp [hunk 7]
```diff
@@ -772,23 +773,446 @@ namespace quantum_sync{
...
+    * @tparam RealType the floating point type to be used
+    * @tparam VectorType the vector type to be used internally (e.g. std::vector)
+    */
+    template<typename RealType, typename VectorType>
+    class tailopt_photon_emission_lookup_table
+    {
+        public:
+
+            /**
+            * A view_type is essentially a photon_emission_lookup_table which
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables.hpp]:
```suggestion
            * A view_type is essentially a tailopt_photon_emission_lookup_table which
```

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables.hpp [hunk 8]
```diff
@@ -772,23 +773,446 @@ namespace quantum_sync{
...
+                        math::m_log(math::one<RealType>),
+                        math::m_log(m_params.frac_switch))}}
+                {}
+
+            /**
+            * Constructor (not designed for GPU usage)
+            * This constructor allows the user to initialize the table with
+            * a vector of values.
+            *
+            * @param params table parameters
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables.hpp:939]:
```suggestion
            * @param params table parameters
              @param vals the values used to initialize the table
```

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables.hpp [hunk 9]
```diff
@@ -772,23 +773,446 @@ namespace quantum_sync{
...
+            * If X is out of table, 0 is returned (i.e. a photon with
+            * very low energy is emitted, so it can be disregarded)
+            * The method also checks if chi_phot is out of table
+            * and stores the result in a bool variable.
+            *
+            * @param[in] chi_phot to be used for interpolation
+            * @param[in] unf_zero_one_minus_epsi a uniformly distributed random number in [0,1)
+            * @param[out] is_out set to true if chi_part is out of table
+            *
+            * @return chi of one of the generated particles
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables.hpp]:
```suggestion
            * @return chi of the generated photon```

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables.hpp [hunk 10]
```diff
@@ -772,23 +773,446 @@ namespace quantum_sync{
...
+            /**
+            * Imports table values from an std::vector. Values
+            * should correspond to coordinates exported with
+            * get_all_coordinates(). Not usable on GPU.
+            *
+            * @param[in] a std::vector containing table values
+            *
+            * @return false if the value vector has the wrong length. True otherwise.
+            */
+            bool set_all_vals(const std::vector<RealType>& vals)
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables.hpp:1155]: You may want to use the new method that updates all the values at once, which was added in #21?

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_detail.hpp [hunk 11]
```diff
@@ -0,0 +1,513 @@
...
+            serialization::put_in(m_zmin, raw);
+            serialization::put_in(m_coeff, raw);
+
+            return raw;
+        }
+
+        /**
+        * Deserializes a LinFunctor object from binary data (not suitable for GPUs)
+        *
+        * @tparam CharIter a char interator type
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_detail.hpp]:
```suggestion
        * @tparam CharIter a char iterator type
```

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_detail.hpp [hunk 12]
```diff
@@ -0,0 +1,513 @@
...
+        *
+        * @param[in,out] it a char interator
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_detail.hpp]:
```suggestion
        * @param[in,out] it a char iterator
```

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_detail.hpp [hunk 13]
```diff
@@ -0,0 +1,513 @@
...
+        }
+
+        /**
+        * Deserializes a TailOptFunctor object from binary data (not suitable for GPUs)
+        *
+        * @tparam CharIter a char interator type
+        *
+        * @param[in,out] it a char interator
+        *
+        * @return a LinFunctor<RealType> object
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_detail.hpp]:
```suggestion
        * @return a TailOptFunctor<RealType> object
```

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_detail.hpp [hunk 14]
```diff
@@ -0,0 +1,513 @@
...
+        {
+            m_exp_zswitch = math::m_exp(zswitch);
+            m_coeff_first = (zfirst - 1) / (zswitch - zmin);
+            m_coeff_second = (zsize - zfirst) / (math::m_exp(zmax) - m_exp_zswitch);
+        }
+
+        /**
+        * Operator()
+        *
+        * @param[in] i an integer value
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_detail.hpp]:
```suggestion
        * @param[in] z a real value
```

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_detail.hpp [hunk 15]
```diff
@@ -0,0 +1,513 @@
...
+        }
+
+        /**
+        * Deserializes a TailOptFunctor object from binary data (not suitable for GPUs)
+        *
+        * @tparam CharIter a char interator type
+        *
+        * @param[in,out] it a char interator
+        *
+        * @return a LinFunctor<RealType> object
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_detail.hpp]:
```suggestion
        * @return an ITailOptFunctor<RealType> object
```

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_generator_detail.hpp [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+
+    //__________________________________________________________________________
+
+}
+}
+}
+}
+}
+
+#endif // PICSAR_MULTIPHYSICS_QUANTUM_SYNC_ENGINE_TABLES_GENERATOR
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_generator_detail.hpp]:
```suggestion
#endif // PICSAR_MULTIPHYSICS_QUANTUM_SYNC_ENGINE_TABLES_GENERATOR_DETAIL
```

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_generator_detail.hpp [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+//in QED_tests/validation for a more in-depth discussion.
+//
+// References:
+// 1) C.P.Ridgers et al. Journal of Computational Physics 260, 1 (2014)
+// 2) A.Gonoskov et al. Phys. Rev. E 92, 023305 (2015)
+
+//Should be included by all the src files of the library
+#include "picsar_qed/qed_commons.h"
+
+//Uses BW tabulated functions
```
Comment (NeilZaim) [inline multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_generator_detail.hpp]:
```suggestion
//Uses QS tabulated functions
```

**Review by NeilZaim — COMMENTED** [review]
Thanks a lot for this PR! It looks good to me, I only left minor comments.

There's a bit of code duplication between the old tables and the tail optimized tables, but I don't think it's worth the effort of a huge refactoring at this point.

Do you want to add these new tables in WarpX eventually?

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_detail.hpp (hunk 11 above)
Comment (lucafedeli88) [reply multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_detail.hpp]: thanks!

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_generator_detail.hpp (hunk 17 above)
Comment (lucafedeli88) [reply multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables_generator_detail.hpp]: thanks!

File: multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp (hunk 1 above)
Comment (lucafedeli88) [reply multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp]: thanks!

File: multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp (hunk 2 above)
Comment (lucafedeli88) [reply multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp:991]: you are right!

Comment (lucafedeli88) [reply multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp:991]: (see comments)

File: multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp (hunk 3 above)
Comment (lucafedeli88) [reply multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp:1005]: yes! you are right! Thanks for spotting this!

File: multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp (hunk 5 above)
Comment (lucafedeli88) [reply multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp:1313]: yes! Thanks!

File: multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp (hunk 6 above)
Comment (lucafedeli88) [reply multi_physics/QED/include/picsar_qed/containers/picsar_tables.hpp:1331]: Yes! Thanks!

File: multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables.hpp (hunk 10 above)
Comment (lucafedeli88) [reply multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables.hpp:1155]: yes, you are right! I had to implement it though...

Comment (lucafedeli88) [reply multi_physics/QED/include/picsar_qed/physics/quantum_sync/quantum_sync_engine_tables.hpp:1155]: I've also added some tests

**Comment by lucafedeli88** [comment]
Thanks for the review. @NeilZaim ! I should have fixed all the issues...

**Review by NeilZaim — APPROVED** [review]
Thanks for this PR!
